1.11 inventories

gr8pefish opened the pull request:
Add in a _basic_ inventory/container/GUI for backpacks.

Check commit messages for detailed information.

gr8pefish committed:
Misc small changes

Add new ToDo file, and unused BackpackSize class

gr8pefish committed:
Add container/inventory for backpack

Currently hardcoded to 1 inventory for all ItemBackpacks.
ItemBackpack holds the ITEM_HANDLER_CAPABILITY, and deals with registering that
InventoryBackpack implements IItemHandlerModifiable, and acts as a simple wrapper for it
ContainerBackpack extends Container, and lays out the slots

gr8pefish committed:
Add GUI to backpacks

GuiHandler created and registered
GuiBackpack is hardcoded and rudimentary, but works for testing for now

TehNut commented on src/main/java/gr8pefish/ironbackpacks/container/GuiBackpack.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY) {
+        FontRenderer fontRendererObj = Minecraft.getMinecraft().fontRenderer;
```
GuiScreen provides the `fontRenderer` object already.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/container/GuiBackpack.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    protected void drawGuiContainerBackgroundLayer(float partialTicks, int mouseX, int mouseY) {
+        GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+        mc.getTextureManager().bindTexture(texture);
```
I'm going to move the GUI texture stuff to a system [like I use in Nomagi](https://github.com/SleeplessAcorn/Nomagi/blob/1.11/src/main/java/info/sleeplessacorn/nomagi/util/SubTexture.java).

gr8pefish replied:
Alright sounds good. I might leave it as-is for now, and let you handle that part then, as I think it will be easier for you to just knock it out real quick.

TehNut commented on src/main/java/gr8pefish/ironbackpacks/item/ItemBackpack.java:
```diff
@@ -7,20 +7,35 @@
 
 public class ItemBackpack extends Item implements IBackpack {
 
+    public static final int GUI_ID = 0; //ToDo: Where to move this?
```
Should probably be in the GuiHandler

TehNut commented on src/main/java/gr8pefish/ironbackpacks/IronBackpacks.java:
```diff
@@ -44,6 +46,7 @@ public void preInit(FMLPreInitializationEvent event) {
 	@Mod.EventHandler
 	public void init(FMLInitializationEvent event) {
 	    PROXY.init(event);
+		NetworkRegistry.INSTANCE.registerGuiHandler(INSTANCE, new GuiHandler()); //ToDo: Move somewhere else?
```
Probably into `CommonProxy`, just to keep the main class clean.

gr8pefish committed:
Misc fixes/cleanup

Move OPEN_PACK_GUI_ID to GuiHandler
FontRenderer already in GuiClass
Move GuiHandler registration to CommonProxy

TehNut committed:
Fixed nullability in container slot transfer

Signed-off-by: Nicholas Ignoffo <speedynutty68@gmail.com>

TehNut committed:
BackpackInfo now infers it's inventory from the capability

Signed-off-by: Nicholas Ignoffo <speedynutty68@gmail.com>

gr8pefish committed:
More prodding at IItemHandler impl

Still not working, need to deal with registering/accessing itemStack specific variants

gr8pefish committed:
Messy commit, proof dynamic IItemHandler can work

Will cleanup files later, just wanted to have this commit to reference just in case

gr8pefish committed:
Much cleaner inventory implementation

gr8pefish committed:
Major cleanup

gr8pefish commented:
Okay, much improved now, the inventory sizes change dynamically depending on the backpack's stats, are saved to the itemstack, and persist over time (as you'd expect, since it is essentially just a crazy wrapper/implementation of IItemHandler). 

I'd like you to review a fair bit, as I added classes to the API and changed around a bit of code which I imagine you may want to move around/modify yourself.

gr8pefish committed:
Minor cleanup

Remove unused file(s)

gr8pefish commented:
@TehNut Finally back. You want to do any review/changes before I go ahead and merge?

gr8pefish committed:
Start of API refactor

Move to packages
See APIStructure.md for layout

gr8pefish committed:
Start of cleanup on BackpackVariants

gr8pefish committed:
More API refactoring

Inventories don't save, but the rest is coming along nicely (still lots of work to be done)

gr8pefish committed:
Hardcoded VariantEnum working for inventories

Now to reverse engineer incrementally to make the dynamic system functional instead

gr8pefish committed:
More testing of where VariantEnum diverges

Eliminated cause of error as both HashMap and variantList, so that is nice progress at least.

gr8pefish committed:
Change getBackpackSize to be @Nullable

gr8pefish committed:
Dynamic inventories working!

Remove BackpackVariantEnum (just useful for testing)

The result was that the .containsKey()/equality for the HashMap of BackpackVariant wasn't working according to my assumptions, as changing the nbt inventory tag changed the hash and therefore it was no longer "equal". Overriding .equals() and .hash() in BackpackVariant solved this issue.

gr8pefish committed:
BackpackVariant + Misc cleanup

gr8pefish committed:
Minor cleanup/comments

gr8pefish committed:
Major cleanup

Move packages around
Tons of comments added
Couple of methods added and/or moved

gr8pefish committed:
Fix minor issues

Backpack size changing itself too much
Correct import for inventory capability

gr8pefish committed:
Minor method name changes for clarity

Optimize imports too
Must appease TehNut :P

gr8pefish committed:
More method name changes

Verbosity = clarity, which is more useful than succint method names

gr8pefish committed:
Start of dynamic GUI

Some other minor tweaks, plenty of work still needed on this
Thanks @copygirl!

gr8pefish committed:
Dynaimc GUI working with 7 rows now

gr8pefish committed:
Cleanup GUI code

gr8pefish committed:
Quick 3.0.0 alpha build

gr8pefish committed:
Major cleanup

GUI/Container/API, etc.
Lots of preconditions and comments added, and methods moved around.
Reorganized most TODOs as well.
Mostly cleaned up at this point, for what I've gone through, but a couple stragglers here and there still.

TehNut committed:
Fix capability implementation

Signed-off-by: Nicholas Ignoffo <speedynutty68@gmail.com>

TehNut committed:
Cleanup container to not allow moving open backpack

Signed-off-by: Nicholas Ignoffo <speedynutty68@gmail.com>

TehNut commented:
I'm going to go ahead and merge this
